canonical/tempo-k8s-operator
# Add traefik_route to serve tracing endpoints through ingress if it exists

## Issue
Tempo receiver endpoints were not accessible over traefik as they are bound to different ports. Fixes #82 

## Solution
Use `traefik_route` with changes introduced in [traefik-k8s-operator#325](https://github.com/canonical/traefik-k8s-operator/pull/325) to define ports traefik needs to open.
Also fix creating URLs for tracing consumers that use `external_url` - scheme and port were missing from the created url.

## Context
This change is needed for cross-model traces to work. They cannot access internal endpoints of a different k8s cluster . No traces were also coming from machine models.

## Testing Instructions
Run the following bundle and verify endpoints (example, port 3200, `/metrics`, port 4318 (expected prometheus metrics), `/v1/traces` (expected HTTP code 405)) are accessible on traefik's external IP.
```
bundle: kubernetes
applications:
  tempo:
    charm: local:tempo-k8s-1
    scale: 1
    constraints: arch=amd64
    storage:
      data: kubernetes,1,1024M
  traefik-k8s:
    charm: traefik-k8s
    channel: latest/edge
    revision: 184
    base: ubuntu@20.04/stable
    resources:
      traefik-image: 159
    scale: 1
    constraints: arch=amd64
    storage:
      configurations: kubernetes,1,1024M
relations:
- - tempo:tracing
  - traefik-k8s:tracing
- - traefik-k8s:traefik-route
  - tempo:ingress
```


## Release Notes
